src/porepy/numerics/ad/operators.py

```diff
@@ -38,6 +40,62 @@ def _get_shape(mat):
         return mat.shape
 
 
+def _get_previous_time_or_iterate(
+    op: Operator, prev_time: bool = True, steps: int = 1
+) -> Operator:
+    """Helper function which traverses an operator tree by recursion to get a
+    copy of and its childrent, representing it at a previous time or
+    previous iteration.
+
+    Parameters:
+        op: Some operator whos tree should be traversed.
+        prev_time: ``default=True``
+
+            If True, it calles :meth:`Operator.previous_timestep`, otherwise it calls
+            :meth:`Operator.previous_iteration`.
+
+            This is the only difference in the recursion and we can avoid duplicate
+            code.
+        steps: ``default=1``
+
+            Number of steps backwards in time or iterate sense.
+
+    Returns:
+        A copy of the operator and its children, representing the previous time or
+        iterate.
+
+    """
+
+    # The recursion reached an atomic operator, which has some time- or
+    # iterate-dependent behaviour
+    if isinstance(op, TimeDependentOperator) and prev_time:
+        return op.previous_timestep(steps=steps)
+    elif isinstance(op, IterativeOperator) and not prev_time:
+        return op.previous_iteration(steps=steps)
+    # NOTE The previous_iteration of a time-dependent operator will return the operator
```

Comment:
Why is this? And does it hold even for steps > 1?

Reply:
It is steps-independent, yes.
And at this point two things can happen:
The time-dependent operator is original (current time), and it will return itself.
If it is at a previous time, the override in TimeDependentOperator will raise an error, that you cannot take the previous iterate. (But only if the operator inherits from both, time and iterative operator, like the variables do).